utilities/include/ray.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef RAY_HPP
+#define RAY_HPP
+
+#include <iostream>
+#include <cmath>
+#include "vec3.hpp"
+
+float epsilon=0.0001;
+class ray{
+    //origin
+    vec3 origin;
+
+    //unit vector denoting direction
+    vec3 direction;
+
+    public:
+        //Constructors
+        constexpr ray() : origin {vec3(0,0,0)}, direction {vec3(1,0,0)} {}
+        constexpr ray(vec3 Origin, vec3 Direction) : origin {Origin} , direction {Direction.unit()} {}
+
+        //fetch point at distance from point along ray, from origin by default
+        constexpr vec3 fetch(float distance, const vec3& point) const{
+            return point + distance * direction;
+        }
+
+        constexpr vec3 fetch(float distance) const{
+            return origin + distance * direction;
+        }
+
+        //origin setter
+        vec3& Origin(const vec3& Origin){
```

Comment:
I don't think the setters for `origin` and `direction` are required because they are passed in in the constructor itself.